recipe/setup.py

```diff
@@ -132,12 +132,106 @@
+                "csrc/layer_norm/ln_fwd_3072.cu",
+                "csrc/layer_norm/ln_bwd_3072.cu",
+                "csrc/layer_norm/ln_fwd_4096.cu",
+                "csrc/layer_norm/ln_bwd_4096.cu",
+                "csrc/layer_norm/ln_fwd_5120.cu",
+                "csrc/layer_norm/ln_bwd_5120.cu",
+                "csrc/layer_norm/ln_fwd_6144.cu",
+                "csrc/layer_norm/ln_bwd_6144.cu",
+                "csrc/layer_norm/ln_fwd_7168.cu",
+                "csrc/layer_norm/ln_bwd_7168.cu",
+                "csrc/layer_norm/ln_fwd_8192.cu",
+                "csrc/layer_norm/ln_bwd_8192.cu",
+                "csrc/layer_norm/ln_parallel_fwd_256.cu",
+                "csrc/layer_norm/ln_parallel_bwd_256.cu",
+                "csrc/layer_norm/ln_parallel_fwd_512.cu",
+                "csrc/layer_norm/ln_parallel_bwd_512.cu",
+                "csrc/layer_norm/ln_parallel_fwd_768.cu",
+                "csrc/layer_norm/ln_parallel_bwd_768.cu",
+                "csrc/layer_norm/ln_parallel_fwd_1024.cu",
+                "csrc/layer_norm/ln_parallel_bwd_1024.cu",
+                "csrc/layer_norm/ln_parallel_fwd_1280.cu",
+                "csrc/layer_norm/ln_parallel_bwd_1280.cu",
+                "csrc/layer_norm/ln_parallel_fwd_1536.cu",
+                "csrc/layer_norm/ln_parallel_bwd_1536.cu",
+                "csrc/layer_norm/ln_parallel_fwd_2048.cu",
+                "csrc/layer_norm/ln_parallel_bwd_2048.cu",
+                "csrc/layer_norm/ln_parallel_fwd_2560.cu",
+                "csrc/layer_norm/ln_parallel_bwd_2560.cu",
+                "csrc/layer_norm/ln_parallel_fwd_3072.cu",
+                "csrc/layer_norm/ln_parallel_bwd_3072.cu",
+                "csrc/layer_norm/ln_parallel_fwd_4096.cu",
+                "csrc/layer_norm/ln_parallel_bwd_4096.cu",
+                "csrc/layer_norm/ln_parallel_fwd_5120.cu",
+                "csrc/layer_norm/ln_parallel_bwd_5120.cu",
+                "csrc/layer_norm/ln_parallel_fwd_6144.cu",
+                "csrc/layer_norm/ln_parallel_bwd_6144.cu",
+                "csrc/layer_norm/ln_parallel_fwd_7168.cu",
+                "csrc/layer_norm/ln_parallel_bwd_7168.cu",
+                "csrc/layer_norm/ln_parallel_fwd_8192.cu",
+                "csrc/layer_norm/ln_parallel_bwd_8192.cu",
+            ],
+            extra_compile_args={
+                "cxx": ["-O3"],
+                "nvcc": [
+                    "-O3",
+                    "-U__CUDA_NO_HALF_OPERATORS__",
+                    "-U__CUDA_NO_HALF_CONVERSIONS__",
+                    "-U__CUDA_NO_BFLOAT16_OPERATORS__",
+                    "-U__CUDA_NO_BFLOAT16_CONVERSIONS__",
+                    "-U__CUDA_NO_BFLOAT162_OPERATORS__",
+                    "-U__CUDA_NO_BFLOAT162_CONVERSIONS__",
+                    "--expt-relaxed-constexpr",
+                    "--expt-extended-lambda",
+                    "--use_fast_math",
+                ]
+            },
+            extra_link_args = ["-Wl,--strip-all"],
```

Comment:
```suggestion
            },
            libraries=[
                'cublas',
                'cublasLt',
            ],
            extra_link_args = ["-Wl,--strip-all"],
```

https://github.com/Dao-AILab/flash-attention/blob/478ee666cccbd1b8f63648633003059a8dc6827d/csrc/fused_dense_lib/fused_dense_cuda.cu#L11

Reply:
Also [needs cuRAND]( https://github.com/Dao-AILab/flash-attention/blob/478ee666cccbd1b8f63648633003059a8dc6827d/csrc/layer_norm/ln_fwd_kernels.cuh#L10 )

```suggestion
            },
            libraries=[
                'cublas',
                'cublasLt',
                'curand',
            ],
            extra_link_args = ["-Wl,--strip-all"],
```

Reply:
Looks like the original `setup.py` doesn't include these libraries?

https://github.com/Dao-AILab/flash-attention/blob/main/csrc/layer_norm/setup.py

Reply:
Right Daniel is stating they should be based on usage found internally, which also makes sense to me

We can also propose they include this change upstream

Reply:
Sorry not following

The `#include`s Daniel and I reference come from cuBLAS and cuRAND. Meaning the symbols used also come from those libraries

Likely we have gotten lucky as `import torch` causes the loader to find these libraries first and thus satisfy the symbols by the time these extensions use them. However we shouldn't rely on this for at least three reasons:

1. Loading order could change
2. If PyTorch changes its dependencies, we won't get them
3. These packages need to express their version constraint on these libraries so they are correctly satisfied at install time

Reply:
Hmm looks like these libraries are explicitly loaded by pytorch, e.g. https://github.com/pytorch/pytorch/blob/6734cb7bf2c1763118dcc430cee6110a88f8f849/torch/__init__.py#L313, since these packages are all pytorch `CUDAExtension`s, perhaps they should rely on pytorch to load them.

Reply:
Surprisingly, the linker seems to think that none of the curand symbols are needed to be loaded dynamically. Perhaps, this package uses header stuff that can be inlined? I have added the cublas, cudart, and python libraries to the linking as needed.